File: javaee-testing/src/main/java/com/gnefedev/javaee/offline/TestsFilter.java
package com.gnefedev.javaee.offline;

import com.gnefedev.javaee.junit.JavaeeTestRunner;
import org.junit.runner.RunWith;
import org.springframework.core.type.classreading.MetadataReader;
import org.springframework.core.type.filter.AbstractTypeHierarchyTraversingFilter;

/**
 * Created by gerakln on 03.09.16.
 */
class TestsFilter extends AbstractTypeHierarchyTraversingFilter {
    TestsFilter() {
        super(false, false);
    }

    @Override
    protected boolean matchSelf(MetadataReader metadataReader) {
        try {
            String className = metadataReader.getClassMetadata().getClassName();
            return isTestClass(Class.forName(className));
        } catch (ClassNotFoundException e) {
            return false;
        }
    }

    static boolean isTestClass(Class<?> candidateClass) throws ClassNotFoundException {
        if (!candidateClass.isAnnotationPresent(RunWith.class)) {
            return false;
        } else {
            return candidateClass
                    .getAnnotation(RunWith.class)
                    .value()
                    .equals(JavaeeTestRunner.class);
        }
    }
}

File: jee-testing/offline/src/main/java/com/gnefedev/jee/testing/offline/scanner/TestsFilter.kt
package com.gnefedev.jee.testing.offline.scanner

import com.gnefedev.jee.testing.offline.isTest
import org.springframework.core.type.classreading.MetadataReader
import org.springframework.core.type.filter.AbstractTypeHierarchyTraversingFilter

/**
 * Created by gerakln on 03.09.16.
 */
internal class TestsFilter : AbstractTypeHierarchyTraversingFilter(false, false) {
    override fun matchSelf(metadataReader: MetadataReader?): Boolean {
        try {
            val className = metadataReader!!.classMetadata.className
            return Class.forName(className).isTest()
        } catch (e: ClassNotFoundException) {
            return false
        }

    }
}
